src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -776,6 +777,132 @@ bool MetaspaceShared::try_link_class(InstanceKlass* ik, TRAPS) {
+  char* start = buffer + strlen(buffer);
+  snprintf(start, buff_len, "%s ", arg);
+}
+
+void MetaspaceShared::cmd_dump_shared_archive(outputStream* output, const char* cmd, const char* file_name,  TRAPS) {
+  // The existing file will be overwritten.
+  char filename[JVM_MAXPATHLEN];
+  const char* file = file_name;
+  assert(strcmp(cmd, "static_dump") == 0 || strcmp(cmd, "dynamic_dump") == 0, "Sanity check");
+  bool is_static = strcmp(cmd, "static_dump") == 0;
+  if (is_static) {
+    output->print_cr("Static dump");
+    if (file_name ==nullptr) {
+      os::snprintf(filename, sizeof(filename), "java_pid%d_static.jsa", os::current_process_id());
+      file = filename;
+    } else {
+      if (strstr(file_name, ".jsa") == nullptr) {
+        os::snprintf(filename, sizeof(filename), "%s.jsa", file_name);
+        file = filename;
+      }
+    }
+    cmd_dump_static(output, file, THREAD);
+  } else {
+    output->print_cr("Dynamic dump");
+    if (!UseSharedSpaces) {
+      output->print_cr("CDS is not available for this version.");
+      return;
+    }
+    if (!RecordDynamicDumpInfo) {
+      output->print_cr("Please run with -Xshare:auto -XX:+RecordDynamicDumpInfo dumping dynamic archive!");
+      return;
+    }
+    if (file_name == nullptr) {
+      os::snprintf(filename, sizeof(filename), "java_pid%d_dynamic.jsa", os::current_process_id());
+      file = filename;
+    } else {
+      if (strstr(file_name, ".jsa") == nullptr) {
+        os::snprintf(filename, sizeof(filename), "%s.jsa", file_name);
+        file = filename;
+      }
+    }
+    cmd_dump_dynamic(output, file, THREAD);
+  }
+}
+
+class DumpClassListCLDClosure : public CLDClosure {
+  fileStream *_stream;
+public:
+  DumpClassListCLDClosure(fileStream* f) : CLDClosure() { _stream = f; }
+  ~DumpClassListCLDClosure() {
+    delete _stream; // The file need close since in child process it will be used.
+  }
+  void do_cld(ClassLoaderData* cld) {
+    for (Klass* klass = cld->klasses(); klass != NULL; klass = klass->next_link()) {
+      if (klass->is_instance_klass()) {
+        InstanceKlass* ik = InstanceKlass::cast(klass);
+        if (ik->is_shareable()) {
+          _stream->print_cr("%s", ik->name()->as_C_string());
+        }
+      }
+    }
+  }
+};
+
+//   To create a static dump, steps:
+//   1. output a classlist file
+//   2. fork a new process to dump the shared archive
+void MetaspaceShared::cmd_dump_static(outputStream* output, const char* file, TRAPS) {
+  const char* java_home = Arguments::get_java_home();
+  const char* file_separator = os::file_separator();
+  const char* app_class_path = Arguments::get_appclasspath();  // -cp ..
+  const char* java_command = Arguments::java_command();        // arguments to app
+  char exec_path[JVM_MAXPATHLEN]; // $JAVA_HOME/bin/java or %JAVA_HOME%\\jre\bin\\java ...
+  char classlist_name[JVM_MAXPATHLEN];
+
+  os::snprintf(classlist_name, sizeof(classlist_name), "%s.classlist", file);
```

Comment:
I think the `file` contains the ".jsa" suffix. So the `classlist_name` would be <name>.jsa.classlist.
Maybe only the prefix of `file` should be passed into cmd_dump_static() and cmd_dump_dynamic() and have the functions append the suffix for the names for the classlist and archive?